### PR TITLE
Add Inline YAML Program Functionality

### LINKED ISSUE #312: Support inline programs
At present, the mode of using this operator is to create a Pulumi project in a git repo, then refer to that git repo in a Stack object. This works fine, especially if you use the Pulumi project for other deployments.

But it is not the _simplest_ useful way it could work: indirection through a git repo is a complication if you just want to automate your infrastructure with the operator.

To serve that use case, the operator could run programs supplied within a Kubernetes resource.

Anticipated questions:

**But is it GitOps?**

This is still amenable to GitOps, because you can manage the resource definitions (i.e., the program) with GitOps tooling like Flux -- i.e., keep them in files in git, and get Flux to sync them to the cluster. For some people, having the program encapsulated in a resource and syncing the resource will be more natural than referring to code in a git repo.

**Which languages should it support?**

Intuitively, Pulumi YAML seems like the first choice, since it 1. is self-contained (you don't have a bunch of files) and 2. is the same superficial syntax as the enclosing resource. For other languages, the converse of each of 1) and 2) are strong points against -- having to encode e.g., several JavaScript files (and a package.json) within a YAML file would be awkward enough that it's not worth the effort.

Open questions:

 - how to address the mismatch between Pulumi YAML and Kubernetes YAML
 - are programs a different resource type, or an alternative to a Stack source

### PR DESCRIPTION
**Pull request by roothorp**
Fixes #312

### CODE REVIEW

**Comment by squaremo** [comment]
Test cases:

 - [x] Can create the Program value and savee it; and create a Stack that refers to it
 - [x] A valid Program/Stack pair will run the stack (make the Program do something minimal, though)
   - need to save kubeconfig and refer to it in env as KUBECONFIG (examples in other tests) 
 - [ ] A Stack refering to a missing Program will fail and be marked "retrying" (I think there's a SourceUnavailable reason)
 - [ ] A Program that is syntactically OK but invalid (e.g., dangling variable reference) will cause the Stack to fail and be marked stalled, reason: "incorrect program"

File: deploy/crds/pulumi.com_programs.yaml
Comment (lblackstone) [inline deploy/crds/pulumi.com_programs.yaml:22]: Would there be any reason to add this CRD as a pre-v1 version, or are you sure about the API?

File: pkg/controller/stack/stack_controller.go
Comment (lblackstone) [inline pkg/controller/stack/stack_controller.go:317]: I believe @squaremo is refactoring this selection logic in #324 already, but we should make sure to validate any mutually exclusive options rather than implicitly selecting whatever comes first in this list. i.e., the current logic will implicitly use the `GitSource` if multiple options are set.

File: test/program_test.go
Comment (lblackstone) [inline test/program_test.go:1]:
```suggestion
// Copyright 2022, Pulumi Corporation.  All rights reserved.
```

File: test/program_test.go
Comment (lblackstone) [inline test/program_test.go:193]: I think you'll also need the kubeconfig to make sure the failure isn't related to lack of creds. Same for previous test case.

File: pkg/apis/pulumi/shared/stack_types.go
Comment (squaremo) [inline pkg/apis/pulumi/shared/stack_types.go:67]:
```suggestion
	// ProgramRef refers to a Program object, to be used as the source for the stack.
```

File: pkg/controller/stack/stack_controller.go
Comment (squaremo) [reply pkg/controller/stack/stack_controller.go:317]: See https://github.com/pulumi/pulumi-kubernetes-operator/pull/324/commits/fc4f6fce787390d90af098facdad90083e7c4484#diff-6b7314d866c388df2be1614ad056c21202cdcbecace35a4496fb6eeb36ca3dc3R272 and what follows.

(I would simply wait until that's merged, and rewrite this to match when rebasing)

File: pkg/controller/stack/stack_controller.go
Comment (squaremo) [inline pkg/controller/stack/stack_controller.go:808]: Worth an explanatory comment, I think.

File: pkg/controller/stack/stack_controller.go
Comment (squaremo) [inline pkg/controller/stack/stack_controller.go:866]: NB: in #324 I have factored here down out into a helper; watch out for that if you end up rebasing.

File: test/program_test.go
Comment (squaremo) [inline test/program_test.go:131]: Stray debug output?

File: test/program_test.go
Comment (squaremo) [inline test/program_test.go:129]: I think if the stack has succeeded, you should be able to just fetch the configmap straight away (e.g., use `Expect(k8sClient.Get(...)).To(Succeed())`.

File: test/program_test.go
Comment (squaremo) [inline test/program_test.go:138]: Why is this here?

File: test/program_test.go
Comment (squaremo) [inline test/program_test.go:232]: You can unmarshal directly into `&prog.Program` here -- if it fails, you'll throw away the whole value anyway.

File: test/program_test.go
Comment (squaremo) [inline test/program_test.go:234]: You can use `ExpectWithOffset(1, err).ToNot(HaveOccurred())`, which is a bit more consistent and compact. The `WithOffset` bit refers to an offset in the call stack, so that if you are calling this from test cases, the test report highlight the failing line in the test case rather than the failing line here.

File: test/program_test.go
Comment (squaremo) [inline test/program_test.go:222]: :100: for using a helper for this!

**Review by squaremo — CHANGES_REQUESTED** [review]
Please take my comments as suggestions, in general. This is "request changes" because I would like to see comments on the new API fields, so that there are descriptions in the documentation.

File: deploy/crds/pulumi.com_programs.yaml
Comment (roothorp) [reply deploy/crds/pulumi.com_programs.yaml:22]: We talked about this a bit when pairing on it - I think this is fine, but I'm happy to move it if @squaremo thinks it would be wise to do so.

File: test/program_test.go
Comment (roothorp) [reply test/program_test.go]: That was just coping something I shouldn't have - the lines below are deleting the configmap that the program is creating; otherwise repeating the test fails because the configmap is already there. Is there a better way of doing this?

File: deploy/crds/pulumi.com_programs.yaml
Comment (squaremo) [reply deploy/crds/pulumi.com_programs.yaml:22]: There may be precedents indicating otherwise, but I think generally APIs don't get changed once there's a new version. It'd be a bit odd for an API to depend on -- well, refer to -- a less stable API (certainly one of the same group).

The conservative way to do it would be to introduce a new, unstable version of the API (e.g., v2alpha1) and add the Program type, and change the Stack type, in that. As discussed [elsewhere](https://github.com/pulumi/pulumi-kubernetes-operator/pull/324#discussion_r975792786), expedience puts a thumb on the scales.

File: pkg/apis/pulumi/v1/program_types.go
Comment (squaremo) [reply pkg/apis/pulumi/v1/program_types.go]: Ohhhh yes I see, that's annoying. OK let's keep it "correct", and cope with the funny markdown output until we can fix that in the docs generation tool.

File: test/program_test.go
Comment (squaremo) [reply test/program_test.go]: The standard way to deal with this is to remove things in AfterEach, which will run whether or not it completed the test (so is superior to cleaning up inline in the test case). To do that you'll have to wrap the test case in `When` (or `Context`). You could set `DestroyOnFinalize` on the stack, and delete it in AfterEach, to the same effect. Deleting the stack might mean you can have an ApplyEach that works for many test cases, since (presumably) they all create a stack.

File: test/program_test.go
Comment (squaremo) [inline test/program_test.go:138]: Generally all we'll know is that the automation API failed to run it to completion; a `failed` state might be as good as we can do for now.

**Review by squaremo — APPROVED** [review]
Looks good; there's a couple of cosmetic things that could be changed, but neither is a one way door. Nice work! 🍎

**Comment by squaremo** [comment]
NB on failing tests: the failures already fail on default branch. We'll try to roll default branch forward to fix them.

**Pull request merged**
